src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PotentialEditorConfigDataSource.cs

```diff
@@ -0,0 +1,90 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    [Export(typeof(IFileWatchProviderDataSource))]
+    internal class PotentialEditorConfigDataSource : ProjectValueDataSourceBase<IFileWatchProvider>, IFileWatchProviderDataSource
+    {
+        private readonly ConfiguredProject _project;
+
+        private int _version;
+        private IDisposable _projectRuleSourceLink;
+
+        /// <summary>
+        /// The block that receives updates from the active tree provider.
+        /// </summary>
+        private IBroadcastBlock<IProjectVersionedValue<IFileWatchProvider>> _broadcastBlock;
+
+        /// <summary>
+        /// The public facade for the broadcast block.
+        /// </summary>
+        private IReceivableSourceBlock<IProjectVersionedValue<IFileWatchProvider>> _publicBlock;
+
+        [ImportingConstructor]
+        public PotentialEditorConfigDataSource(ConfiguredProject project)
+            : base(project.Services)
+        {
+            _project = project;
+        }
+
+        public override NamedIdentity DataSourceKey { get; } = new NamedIdentity(nameof(PotentialEditorConfigDataSource));
+
+        public override IComparable DataSourceVersion => _version;
+
+        public override IReceivableSourceBlock<IProjectVersionedValue<IFileWatchProvider>> SourceBlock => _publicBlock;
+
+        public FileChangeKinds ChangeKinds => FileChangeKinds.Added | FileChangeKinds.Removed;
+
+        protected override void Initialize()
+        {
+            base.Initialize();
+
+            _projectRuleSourceLink = _project.Services.ProjectSubscription.ProjectRuleSource.SourceBlock.LinkToAsyncAction(ProcessProjectChanged, ruleNames: PotentialEditorConfigFiles.SchemaName);
```

Comment:
I think you'll need to JoinUpstreamDataSources on this ProjectRuleSource, otherwise, if anyone ever blocks on UI thread on you and it needs to switch to the UI thread to complete, it will result in a deadlock. @lifengl?

Reply:
I've add a `JoinUpstreamDataSources(_project.Services.ProjectSubscription.ProjectRuleSource)`.